File: Tutorials/Codes/Advanced_Robotics_Course/Chapter13/Lesson3/Chapter13_Lesson3.cpp
#include <vector>
#include <array>
#include <cmath>

struct Trajectory {
    std::vector<std::array<double, 7>> q;      // example: 7-DOF joint angles
    std::vector<std::array<double, 7>> qd;     // joint velocities
    std::vector<std::array<double, 3>> ee_pos; // end-effector position
};

struct Experiment {
    Trajectory real_traj;
    // plus inputs, initial state, etc.
};

class SimulatorInterface {
public:
    Trajectory rolloutSim(const std::vector<double>& theta,
                          const Experiment& exp) const;
};

std::vector<double> trajectoryFeatures(const Trajectory& traj) {
    // Concatenate mean and std of q, qd, ee_pos
    const std::size_t T = traj.q.size();
    const std::size_t nq = traj.q.front().size();
    const std::size_t ne = traj.ee_pos.front().size();

    std::vector<double> feats;
    feats.reserve(2 * (nq + nq + ne)); // mean and std for each coord

    auto accumulate_mean_std = [&](const auto& seq, std::size_t dim) {
        std::vector<double> mean(dim, 0.0);
        std::vector<double> sq(dim, 0.0);
        for (const auto& x : seq) {
            for (std::size_t i = 0; i < dim; ++i) {
                mean[i] += x[i];
                sq[i]   += x[i] * x[i];
            }
        }
        for (std::size_t i = 0; i < dim; ++i) {
            mean[i] /= static_cast<double>(T);
            double var = sq[i] / static_cast<double>(T) - mean[i] * mean[i];
            double stdv = std::sqrt(std::max(var, 0.0));
            feats.push_back(mean[i]);
            feats.push_back(stdv);
        }
    };

    accumulate_mean_std(traj.q, nq);
    accumulate_mean_std(traj.qd, nq);
    accumulate_mean_std(traj.ee_pos, ne);

    return feats;
}

double featureMismatch(const Trajectory& real_traj,
                       const Trajectory& sim_traj) {
    auto fr = trajectoryFeatures(real_traj);
    auto fs = trajectoryFeatures(sim_traj);
    double cost = 0.0;
    const std::size_t n = fr.size();
    for (std::size_t i = 0; i < n; ++i) {
        double d = fr[i] - fs[i];
        cost += d * d;
    }
    return cost;
}

double calibrationObjective(const std::vector<double>& theta,
                            const std::vector<Experiment>& exps,
                            const SimulatorInterface& sim) {
    double total = 0.0;
    for (const auto& e : exps) {
        Trajectory sim_traj = sim.rolloutSim(theta, e);
        total += featureMismatch(e.real_traj, sim_traj);
    }
    return total;
}
      

File: Tutorials/Codes/Advanced_Robotics_Course/Chapter13/Lesson3/Chapter13_Lesson3.java
import java.util.*;
import java.util.stream.*;

class Trajectory {
    public double[][] q;
    public double[][] qd;
    public double[][] eePos;
}

class Experiment {
    public Trajectory realTraj;
    // plus commands, initial state, etc.
}

interface SimulatorInterface {
    Trajectory rolloutSim(double[] theta, Experiment exp);
}

public class CalibrationManager {

    private final SimulatorInterface sim;
    private final List<Experiment> experiments;
    private final Random rng = new Random(0L);

    public CalibrationManager(SimulatorInterface sim,
                              List<Experiment> experiments) {
        this.sim = sim;
        this.experiments = experiments;
    }

    private double[] sampleTheta() {
        double[] theta = new double[3];
        for (int i = 0; i < 3; ++i) {
            theta[i] = 0.5 + rng.nextDouble(); // [0.5, 1.5]
        }
        return theta;
    }

    private double[] trajectoryFeatures(Trajectory traj) {
        List<double[]> all = new ArrayList<>();
        all.addAll(Arrays.asList(traj.q));
        all.addAll(Arrays.asList(traj.qd));
        all.addAll(Arrays.asList(traj.eePos));

        List<Double> feats = new ArrayList<>();
        int dim = all.get(0).length;
        int T = traj.q.length;

        for (int k = 0; k < all.size(); ++k) {
            double[] seq = all.get(k);
        }

        // Means and stds for each coordinate across q, qd, eePos
        for (double[][] seqGroup : new double[][][]{traj.q, traj.qd, traj.eePos}) {
            int dimGroup = seqGroup[0].length;
            double[] mean = new double[dimGroup];
            double[] sq = new double[dimGroup];
            int Tgroup = seqGroup.length;
            for (int t = 0; t < Tgroup; ++t) {
                double[] x = seqGroup[t];
                for (int i = 0; i < dimGroup; ++i) {
                    mean[i] += x[i];
                    sq[i] += x[i] * x[i];
                }
            }
            for (int i = 0; i < dimGroup; ++i) {
                mean[i] /= Tgroup;
                double var = sq[i] / Tgroup - mean[i] * mean[i];
                double stdv = Math.sqrt(Math.max(var, 0.0));
                feats.add(mean[i]);
                feats.add(stdv);
            }
        }

        return feats.stream().mapToDouble(d -> d).toArray();
    }

    private double featureMismatch(Trajectory realTraj,
                                   Trajectory simTraj) {
        double[] fr = trajectoryFeatures(realTraj);
        double[] fs = trajectoryFeatures(simTraj);
        double cost = 0.0;
        for (int i = 0; i < fr.length; ++i) {
            double d = fr[i] - fs[i];
            cost += d * d;
        }
        return cost;
    }

    private double calibrationObjective(double[] theta) {
        double total = 0.0;
        for (Experiment e : experiments) {
            Trajectory simTraj = sim.rolloutSim(theta, e);
            total += featureMismatch(e.realTraj, simTraj);
        }
        return total;
    }

    public double[] randomSearch(int numSamples) {
        double bestCost = Double.POSITIVE_INFINITY;
        double[] bestTheta = null;
        for (int k = 0; k < numSamples; ++k) {
            double[] theta = sampleTheta();
            double cost = calibrationObjective(theta);
            if (cost < bestCost) {
                bestCost = cost;
                bestTheta = theta;
            }
        }
        return bestTheta;
    }
}
      
